daemon/containerd/image_exporter.go

```diff
@@ -58,49 +59,95 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, outStrea
-		// Export only those manifests that we have.
-		// TODO(vvoland): Reconsider this when `--platform` is added.
-		if containerdimages.IsIndexType(target.MediaType) {
-			desc, err := i.getBestDescriptorForExport(ctx, target)
-			if err != nil {
-				return err
+		var cleanup func()
+		if cleanup, err = i.leaseContent(ctx, target, func() error {
+			// We may not have locally all the platforms that are specified in the index.
+			// Export only those manifests that we have.
+			// TODO(vvoland): Reconsider this when `--platform` is added.
+			if containerdimages.IsIndexType(target.MediaType) {
+				desc, err := i.getBestDescriptorForExport(ctx, target)
+				if err != nil {
+					return err
+				}
+				target = desc
 			}
-			target = desc
-		}
 
-		if ref, err := reference.ParseNormalizedNamed(name); err == nil {
-			ref = reference.TagNameOnly(ref)
-			opts = append(opts, archive.WithManifest(target, ref.String()))
+			if ref, err := reference.ParseNormalizedNamed(name); err == nil {
+				ref = reference.TagNameOnly(ref)
+				opts = append(opts, archive.WithManifest(target, ref.String()))
 
-			log.G(ctx).WithFields(logrus.Fields{
-				"target": target,
-				"name":   ref.String(),
-			}).Debug("export image")
-		} else {
-			opts = append(opts, archive.WithManifest(target))
+				log.G(ctx).WithFields(logrus.Fields{
+					"target": target,
+					"name":   ref.String(),
+				}).Debug("export image")
+			} else {
+				opts = append(opts, archive.WithManifest(target))
 
-			log.G(ctx).WithFields(logrus.Fields{
-				"target": target,
-			}).Debug("export image without name")
+				log.G(ctx).WithFields(logrus.Fields{
+					"target": target,
+				}).Debug("export image without name")
+			}
+			return nil
+		}); err != nil {
+			cleanup()
+			return err
 		}
+
+		defer cleanup()
 	}
 
 	return i.client.Export(ctx, outStream, opts...)
 }
 
+func (i *ImageService) leaseContent(ctx context.Context, desc ocispec.Descriptor, handler func() error) (func(), error) {
+	ls := i.client.LeasesService()
+	store := i.client.ContentStore()
+	l, err := ls.Create(ctx, leases.WithRandomID())
+	if err != nil {
+		return func() {}, err
+	}
+	cleanup := func() {
+		if err := ls.Delete(ctx, l); err != nil {
+			log.G(ctx).WithError(err).Warn("cleaning up lease")
+		}
+	}
+
+	if err := containerdimages.Walk(ctx, containerdimages.HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		r := leases.Resource{
+			ID:   desc.Digest.String(),
+			Type: "content",
+		}
+		if err := ls.AddResource(ctx, l, r); err != nil {
+			return nil, err
+		}
+
+		_, err := store.Info(ctx, desc.Digest)
+		if err != nil {
+			if errors.Is(err, cerrdefs.ErrNotFound) {
+				return nil, nil
+			}
+			return nil, err
+		}
+
+		c, err := containerdimages.Children(ctx, store, desc)
+		if err != nil {
+			return nil, err
+		}
+
+		return c, nil
+	}), desc); err != nil {
+		return cleanup, err
+	}
```

Comment:
Would it make sense to call cleanup here before returning (in the error case), and in that case return `nil, err`? And the caller only calling `cleanup()` in the non-nil error case?

(same for the other non-nil error returns)

Reply:
Thinking here; if we failed to get a lease (and content), the caller wouldn't have to clean up things, and only clean up ("after I'm done") if we did successfully get a lease.

Reply:
Return `cleanup, handler()` below, and the handler could also return an error so the caller would need to call the cleanup in that case to. Returning in all the time makes it so that the caller has to do the same thing no matter the value of `err`

Reply:
Ok I'm slow this evening, changed it so that the caller only has to call cleanup if there were no errors